src/main/java/org/jenkinsci/plugins/workflow/steps/CoreStep.java

```diff
@@ -75,9 +78,21 @@
         }
 
         @Override protected Void run() throws Exception {
-            FilePath workspace = getContext().get(FilePath.class);
+            final StepContext ctx = this.getContext();
+            final FilePath workspace = Objects.requireNonNull(ctx.get(FilePath.class));
             workspace.mkdirs();
-            delegate.perform(getContext().get(Run.class), workspace, getContext().get(Launcher.class), getContext().get(TaskListener.class));
+            final Run<?,?> run = Objects.requireNonNull(ctx.get(Run.class));
+            final Launcher launcher = Objects.requireNonNull(ctx.get(Launcher.class));
+            final TaskListener listener = Objects.requireNonNull(ctx.get(TaskListener.class));
+            // TODO: Replace with a direct call to SimpleBuildStep.perform(Run, FilePath, EnvVars, Launcher, TaskListener) once the minimum core version for this plugin is 2.241 or newer.
+            try {
+                final Method perform = this.delegate.getClass().getMethod("perform", Run.class, FilePath.class,
+                        EnvVars.class, Launcher.class, TaskListener.class);
+                final EnvVars env = Objects.requireNonNull(ctx.get(EnvVars.class));
+                perform.invoke(this.delegate, run, workspace, env, launcher, listener);
```

Comment:
I prefer to check for the method directly instead of having an indirect check against the Jenkins version. Notably though, this means that the new API would be active in Pipelines even when running old versions of Jenkins core if a `SimpleBuildStep` implements the new method (as seen in the newly-added test case, which now runs against old versions of Jenkins), which could create some confusion.

Reply:
Seems fine to me. If we're running on 2.241 or later, that'll be expected behaviour. And if running on older Jenkins, that other plugin would have had to implement the new API without pulling in the corresponding declarations, which feels like a bit of an edge case.